core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -70,83 +210,147 @@ sealed abstract class Pull[+F[_], +O, +R] {
+    * useful to override compiler type inference.
+    */
+  def covary[G[x] >: F[x]]: Pull[G, O, R] = this
 
-  /** Lifts this pull to the specified output type. */
+  /** Short-hand for `(this: Pull[F, O2, R])`
+    * useful to override compiler type inference.
+    */
   def covaryOutput[O2 >: O]: Pull[F, O2, R] = this
 
-  /** Lifts this pull to the specified resource type. */
-  def covaryResource[R2 >: R]: Pull[F, O, R2] = this
+  /** Short-hand for `(this: Pull[F, O, R2])`
+    * useful to override compiler type inference.
+    */
+  def covaryResult[R2 >: R]: Pull[F, O, R2] = this
+
+  /** Returns a pull with the result wrapped in `Right`, or an error wrapped in `Left` if the pull has failed. */
+  def attempt: Pull[F, O, Either[Throwable, R]] =
+    map(r => Right(r)).handleErrorWith(t => Succeeded(Left(t)))
 
-  /** Applies the resource of this pull to `f` and returns the result in a new `Pull`. */
+  /** Transforms the carry-over result type of this pull to `f` and returns the result in a new `Pull`.
+    *
+    * If `this` pull completes successfully and returns a value `r` of type `R`, 
+    * the function `f` is applied to that result, _after_ this pull has completed
+    * its compilation.
+    */
   def map[R2](f: R => R2): Pull[F, O, R2] =
     new Bind[F, O, R, R2](this) {
       def cont(r: Terminal[R]) = r.map(f)
     }
 
-  /** Run `p2` after `this`, regardless of errors during `this`, then reraise any errors encountered during `this`. */
-  def onComplete[F2[x] >: F[x], O2 >: O, R2 >: R](p2: => Pull[F2, O2, R2]): Pull[F2, O2, R2] =
-    handleErrorWith(e => p2 >> Fail(e)) >> p2
+  /** Discards the result type of this pull.
+    *
+    * Alias for `this.map(_ => () )`.
+    *
+    * This is equivalent to `Functor[Pull[F, O, *]].void(this)`
+    */
+  def void: Pull[F, O, Unit] = as(())
 
-  /** If `this` terminates with `Pull.raiseError(e)`, invoke `h(e)`. */
-  def handleErrorWith[F2[x] >: F[x], O2 >: O, R2 >: R](
-      h: Throwable => Pull[F2, O2, R2]
-  ): Pull[F2, O2, R2] =
-    new Bind[F2, O2, R2, R2](this) {
-      def cont(e: Terminal[R2]): Pull[F2, O2, R2] =
-        e match {
-          case Fail(e) =>
-            try h(e)
-            catch { case NonFatal(e) => Fail(e) }
-          case other => other
-        }
-    }
+  /** Alias for `_.map(_ => o2)`.
+    *
+    * This is equivalent to `Functor[Pull[F, O, *]].as(this, constant)`
+    *
+    * @tparam S     The type of the constant,
+    * @param  value the new result value of the pull
+    */
+  def as[S](value: S): Pull[F, O, S] = map(_ => value)
 
-  /** Discards the result type of this pull. */
-  def void: Pull[F, O, Unit] = as(())
 }
 
 object Pull extends PullLowPriority {
 
+  private[this] val unit: Terminal[Unit] = Succeeded(())
+
+  /** The `done` pull is a pure, closed, silent pull. It succeeds by emitting a `Unit`. */
+  val done: Pull[Pure, INothing, Unit] = unit
+
+  /** Lifts a constant result value into a pure silent Pull that succeesfully
```

Comment:
"constant result value" doesn't say much to me. I'd also drop the reference to applicative. I'd rewrite this to:

```
Creates a pull which evaluates no effects, outputs no elements, and terminates with the supplied value.
```